### ISSUE TITLE
Error: Could not instrument: ..... Unrecognized expression

### ISSUE THREAD

**Issue by gitpusha** [solidity-parser/parser#14]
@fvictorio and @cgewecke thanks for the update in issue #9 .

However, now I get a new Parser Error, I assume one of the following things could be the issue with my expressions in the contract (that compiles fine):

- `pragma experimental ABIEncoderV2` in order to accept solidity `structs` and `nested structs` as `function params` and to `emit` them in `events`. (btw the EncoderV2 is not considered experimental any more since 0.6)

- I am using solidity `try/catch` expressions

```
An unexpected error occurred:

Error: Could not instrument: gelato_core/GelatoCore.sol. (Please verify solc can compile this file without errors.) Unrecognized expression
    at ASTBuilder.Expression (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:758:11)
    at ASTBuilder.visit (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:1304:38)
    at ASTBuilder.Expression (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:700:30)
    at ASTBuilder.visit (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:1304:38)
    at ASTBuilder.TryStatement (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:483:24)
    at ASTBuilder.visit (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:1304:38)
    at ASTBuilder.Statement (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:557:17)
    at ASTBuilder.visit (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:1304:38)
    at ASTBuilder.<anonymous> (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:1292:19)
    at Array.map (<anonymous>)
    at ASTBuilder.visit (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:1291:16)
    at ASTBuilder.Block (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:517:24)
    at ASTBuilder.visit (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:1304:38)
    at ASTBuilder.FunctionDefinition (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:143:20)
    at ASTBuilder.visit (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:1304:38)
    at ASTBuilder.ContractPart (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:127:17)
    at ASTBuilder.visit (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:1304:38)
    at ASTBuilder.<anonymous> (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:1292:19)
    at Array.map (<anonymous>)
    at ASTBuilder.visit (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:1291:16)
    at ASTBuilder.ContractDefinition (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:113:22)
    at ASTBuilder.visit (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-parser-diligence/dist/ASTBuilder.js:1304:38)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by fvictorio** [comment]
@gitpusha can you share a link to the contract that is causing this? Or, if that's not possible, to add a minimal snippet that reproduces the issue?

**Comment by cgewecke** [comment]
@gitpusha Have you manually changed the parser yourself? I have not published a new coverage version yet.

**Comment by gitpusha** [comment]
HI @fvictorio and @cgewecke ,  I was under the impression that there was a new `solidity-coverage` version out since I ran `yarn outdated` and it told me there was a new upgrade `0.7.3`.

Maybe I am on the wrong `solidity coverage` version because I use `0.7.0-beta.3` as was prescribed in the [docs](https://buidler.dev/plugins/solidity-coverage.html#installation) 

So no, I did not update the parser manually yet.

@fvictorio I could send you the whole file, since it's hard to locate the snippet for me. But I think that might be unnecessary now, due to my misunderstanding that the solidity-coverage `0.7.3` would already have the parser update, sorry.

**Comment by cgewecke** [comment]
@gitpusha The parser update will be in `0.7.4`

**Comment by gitpusha** [comment]
Hi @cgewecke and @fvictorio I am just tried out `0.7.4`: 
The good news: Almost all contracts I want covered are successfully instrumented for coverage.
The bad news: At least one of them isnt due to a parser error (see bottom).

 **I assume that the Parser throws this error because it does not recognize Contract Types as keys for solidity mappings.**  See the source code (lines 25 -32):
The error is on line 30: ` mapping (IGelatoProviderModule => uint256) index;`

@fvictorio any chance you can fix this? Thank you both for your help! 

```
library ProviderModuleSet {

    struct Set {
        // Position of the module in the `modules` array, plus 1 because index 0
        // means a module is not in the set.
       mapping (IGelatoProviderModule => uint256) index;
        IGelatoProviderModule[] modules;
    }
....
```

```
ParserError: Could not instrument: libraries/ProviderModuleSet.sol. (Please verify solc can compile this file without errors.) mismatched input 'IGelatoProviderModule' expecting {'address', 'var', 'bool', 'string', 'byte', Int, Uint, Byte, Fixed, Ufixed} (30:17)
    at Object.parse (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/@solidity-parser/parser/dist/index.js:79:11)
    at Instrumenter.instrument (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-coverage/lib/instrumenter.js:63:30)
    at API.instrument (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-coverage/lib/api.js:91:48)
    at SimpleTaskDefinition.action (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/solidity-coverage/plugins/buidler.plugin.js:86:23)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at Environment._runTaskDefinition (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/@nomiclabs/buidler/src/internal/core/runtime-environment.ts:190:14)
    at main (/Users/luis/Desktop/gelato-gnosis-safe/node_modules/@nomiclabs/buidler/src/internal/cli/cli.ts:151:5)
```

**Comment by fvictorio** [comment]
Thanks @gitpusha for the kind words :slightly_smiling_face:

This seems like an easy problem to fix, thanks for letting me know.

**Comment by fvictorio** [comment]
Fixed in 0.6.0.

**Comment by gitpusha** [comment]
Hi @fvictorio . Thank you again for your swift aid!

We just started using the new `immutable` syntax available since `solidity 0.6.5`.

Does the parser already support that syntax?

@cgewecke solidity-coverage `0.7.4` was not able to instrument my contracts that use `immutable`. 

I know the solidity update was only like 10 days ago, but do you plan to upgrade your tools soon?

I am asking because we are currently freezing our codebase that uses all this new syntax for an audit and would be amazing to still have your splendid tools available for all our code now.

**Comment by cgewecke** [comment]
@gitpusha Yes, no problem - I will publish again today.

**Comment by cgewecke** [comment]
@gitpusha Oh I'm sorry - I'm not sure what the status of the immutable syntax is.....

**Comment by fvictorio** [comment]
It should support it, but the AST doesn't indicate that a state variable is immutable. I opened #17 to track that.

**Comment by gitpusha** [comment]
Thanks you both!
 
@fvictorio @cgewecke  I get these when running `solidity coverage 0.7.4` for my contracts that use `immutable`:

**example using immutable combined with override**

``` solidity
address public immutable override gelatoCore;
```
```
ParserError: Could not instrument: user_proxies/gelato_user_proxy/GelatoUserProxyFactory.sol. (Please verify solc can compile this file without errors.) mismatched input 'override' expecting {';', '='} (14:29)
    at Object.parse (/Users/luis/Desktop/gelato-V1/node_modules/@solidity-parser/parser/dist/index.js:79:11)
    at Instrumenter.instrument (/Users/luis/Desktop/gelato-V1/node_modules/solidity-coverage/lib/instrumenter.js:63:30)
    at API.instrument (/Users/luis/Desktop/gelato-V1/node_modules/solidity-coverage/lib/api.js:91:48)
    at SimpleTaskDefinition.action (/Users/luis/Desktop/gelato-V1/node_modules/solidity-coverage/plugins/buidler.plugin.js:86:23)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at Environment._runTaskDefinition (/Users/luis/Desktop/gelato-V1/node_modules/@nomiclabs/buidler/src/internal/core/runtime-environment.ts:190:14)
    at main (/Users/luis/Desktop/gelato-V1/node_modules/@nomiclabs/buidler/src/internal/cli/cli.ts:151:5)
```

**another Example using a Solidity interface**

``` solidity
IBatchExchange private immutable batchExchange;
```
```
ParserError: Could not instrument: gelato_actions/gnosis/ActionPlaceOrderBatchExchange.sol. (Please verify solc can compile this file without errors.) extraneous input 'batchExchange' expecting {';', '='} (17:37)
    at Object.parse (/Users/luis/Desktop/gelato-V1/node_modules/@solidity-parser/parser/dist/index.js:79:11)
    at Instrumenter.instrument (/Users/luis/Desktop/gelato-V1/node_modules/solidity-coverage/lib/instrumenter.js:63:30)
    at API.instrument (/Users/luis/Desktop/gelato-V1/node_modules/solidity-coverage/lib/api.js:91:48)
    at SimpleTaskDefinition.action (/Users/luis/Desktop/gelato-V1/node_modules/solidity-coverage/plugins/buidler.plugin.js:86:23)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at Environment._runTaskDefinition (/Users/luis/Desktop/gelato-V1/node_modules/@nomiclabs/buidler/src/internal/core/runtime-environment.ts:190:14)
    at main (/Users/luis/Desktop/gelato-V1/node_modules/@nomiclabs/buidler/src/internal/cli/cli.ts:151:5)
```

**Comment by cgewecke** [comment]
@gitpusha If you have a chance could you install `solidity-coverage` from the git master branch  and see if all the parsing problems in your project are resolved? (Haven't published 0.7.5 yet...)

```
npm install --save-dev https://github.com/sc-forks/solidity-coverage.git
```

**Comment by gitpusha** [comment]
Hi @cgewecke.  I installed `solidity-coverage` from git master branch and all the parsing problems are resolved ! 

Thank you both very much @fvictorio !
